README.md
# proofservice-weighing-scale
Proxy service for working with weighing scales (only HID-compatible for now)

How-to build
------------
#### Prereqs
 * libusb-1.0 (libusb-1.0-0-dev package in debian-based distributives)

#### Build process
$PROOF_PATH is compulsory for both build and run-time. It should point to directory where Proof binaries and includes (i.e. artifacts of Proof compilation) are placed.

Set $QMAKEFEATURES var to $PROOF_PATH/features dir (needed only for build itself).

Build process:
```bash
mkdir build
cd build
qmake ../proofservice-weighing-scale.pro
make -jN
```
Where N is number of compile tasks that will run in parallel.

Also proof building scripts can be used to build Debian package. proof and proof-dev packages should be installed.

```bash
/opt/Opensoft/proof/tools/build-deb-package -f Manifest
```

Package will be placed to the project root.

How-to use
----------
Config changes are applied only at service restart. If service was installed using deb package then auto restarter should be enabled and simple kill -9 SERVICE_PID will work.

Usually scale device will be created with root-only access. To fix it add file `/etc/udev/rules.d/50-ultegra-scale.rules` (name can be different) with next content (change vendor and product ids to relevant ones):
```
SUBSYSTEMS=="usb", ATTRS{idVendor}=="0b67", ATTRS{idProduct}=="555e", GROUP="dialout", MODE="0666"
```

#### Config example
```ini
[General]
# HID vendor and product ids (can be retrieved from lsusb output)
vendor_id=0x0b67
product_id=0x555e

[server]
# server port
port=9000

[error_notifier]
# app_id which will be added to all emails and system/status or system/recent-errors endpoints
# doesn't do anything else, is used only to separate one instance from another
app_id=DotG-Shipping-Area-Weighing-Scale
# Email section describes credentials for email-based error notifications
# If not enabled nothing will be sent
email\enabled=false
email\from=
email\host=
email\password=
email\port=25
email\to=
# possible values are plain, ssl and starttls
email\type=ssl
email\username=

[updates]
# Should service try to update itself from time to time or not using apt-get(apt-get update && apt-get install)
auto_update=false
# files with sources that should be apt-get updated. If empty all sources will be updated
sources_list_file=
```


CHANGELOG.md
Changelog
=========

## Not Released
#### Features
 * --

#### Bug Fixing
 * --

## 0.18.3.14
Initial release


main.cpp
#include "proofservice-weighing-scale_global.h"
#include "weighingscalerestserver.h"

#include "proofcore/coreapplication.h"
#include "proofcore/settings.h"
#include "proofcore/settingsgroup.h"

#include "proofcore/updatemanager.h"

#include <QTimer>

Q_LOGGING_CATEGORY(proofServiceWeighingScaleLog, "proofservices.weighing-scale")

//TODO: make it cross-platform
int main(int argc, char *argv[])
{
    Proof::CoreApplication a(argc, argv, "Opensoft", "proofservice-weighing-scale", APP_VERSION);

    Proof::SettingsGroup *serverGroup = a.settings()->group("server", Proof::Settings::NotFoundPolicy::Add);
    int serverPort = serverGroup->value("port", 9000, Proof::Settings::NotFoundPolicy::Add).toInt();
    WeighingScaleRestServer server(serverPort);
    server.startListen();

    QObject::connect(a.updateManager(), &Proof::UpdateManager::updateSucceeded, &a, &QCoreApplication::quit);
    QTimer::singleShot(1, &a, &Proof::CoreApplication::postInit);

    return a.exec();
}


weighingscalehandler.h
#ifndef WEIGHINGSCALEHANDLER_H
#define WEIGHINGSCALEHANDLER_H

#include "3rdparty/hidapi/hidapi.h"

#include <QThread>
#include <QReadWriteLock>
#include <QLinkedList>
#include <QTime>

#include <atomic>
#include <functional>

class WeighingScaleHandler : public QThread
{
    Q_OBJECT
public:
    enum class Status {
        UnknownStatus = 0,
        Fault = 1,
        StableZero = 2,
        InMotion = 3,
        Stable = 4,
        UnderZero = 5,
        Overweight = 6,
        DummyValue = 7, // Should never happen, added to make it easier for boundaries check
        NotInitialized = 8
    };
    Q_ENUM(WeighingScaleHandler::Status)

    enum class Unit {
        UnknownUnit = 0,
        Milligram = 1,
        Gram = 2,
        Kilogram = 3,
        Carat = 4,
        Tael = 5,
        Grain = 6,
        Pennyweight = 7,
        MetricTon = 8,
        AvoirTon = 9,
        TroyOunce = 0xA,
        Ounce = 0xB,
        Pound = 0xC
    };
    Q_ENUM(WeighingScaleHandler::Unit)

    struct State {
        QString stringifiedStatus() const;
        QString stringifiedUnit() const;
        Status status;
        Unit unit;
        double weight;
    };

    explicit WeighingScaleHandler(unsigned short vendorId, unsigned short productId, QObject *parent = nullptr);

    State lastStableState() const;
    State instantState() const;
    void execOnNextStableWeight(std::function<void(State)> &&handler);

    bool isAlive() const;
    int elapsedSinceLastMessage() const;
    unsigned short vendorId() const;
    unsigned short productId() const;

    void stop();

protected:
    void run() override;

private:
    unsigned long long packState(Status status, Unit unit, short weight, char scaleFactor) const;
    State extractState(unsigned long long state) const;
    Status extractStateStatus(unsigned long long state) const;

    unsigned short m_vendorId = 0;
    unsigned short m_productId = 0;

    bool m_stopped = false;

    std::atomic_ullong m_instantState {0};
    std::atomic_ullong m_lastStableState {0};

    QLinkedList<std::function<void(State)>> m_stableWaiters;
    QReadWriteLock m_stableWaitersLock;

    hid_device *m_hidHandle = nullptr;
    QTime m_lastSuccessfulRead;
};

uint qHash(WeighingScaleHandler::Unit value, uint seed = 0);
uint qHash(WeighingScaleHandler::Status value, uint seed = 0);

#endif // WEIGHINGSCALEHANDLER_H


weighingscalehandler.cpp
#include "weighingscalehandler.h"

#include "proofservice-weighing-scale_global.h"

#include "proofcore/proofobject.h"

#include <math.h>

constexpr int READ_TIMEOUT = 500;
constexpr int READ_ERROR_TIMEOUT = 5000;
constexpr int ERROR_COOLDOWN_TIMEOUT = 100;

static const QHash<WeighingScaleHandler::Unit, QString> UNITS = {
    {WeighingScaleHandler::Unit::UnknownUnit, ""},
    {WeighingScaleHandler::Unit::Milligram, "mg"},
    {WeighingScaleHandler::Unit::Gram, "g"},
    {WeighingScaleHandler::Unit::Kilogram, "kg"},
    {WeighingScaleHandler::Unit::Carat, "ct"},
    {WeighingScaleHandler::Unit::Tael, "tael"},
    {WeighingScaleHandler::Unit::Grain, "gr"},
    {WeighingScaleHandler::Unit::Pennyweight, "dwt"},
    {WeighingScaleHandler::Unit::MetricTon, "ton"},
    {WeighingScaleHandler::Unit::AvoirTon, "short ton"},
    {WeighingScaleHandler::Unit::TroyOunce, "oz t"},
    {WeighingScaleHandler::Unit::Ounce, "oz"},
    {WeighingScaleHandler::Unit::Pound, "lb"}
};

static const QHash<WeighingScaleHandler::Status, QString> STATUSES = {
    {WeighingScaleHandler::Status::UnknownStatus, "error"},
    {WeighingScaleHandler::Status::Fault, "error"},
    {WeighingScaleHandler::Status::StableZero, "stable"},
    {WeighingScaleHandler::Status::InMotion, "in motion"},
    {WeighingScaleHandler::Status::Stable, "stable"},
    {WeighingScaleHandler::Status::UnderZero, "under zero"},
    {WeighingScaleHandler::Status::Overweight, "overweight"},
    {WeighingScaleHandler::Status::DummyValue, "error"},
    {WeighingScaleHandler::Status::NotInitialized, "not initialized"}
};

uint qHash(WeighingScaleHandler::Unit value, uint seed)
{
    return qHash(static_cast<int>(value), seed);
}

uint qHash(WeighingScaleHandler::Status value, uint seed)
{
    return qHash(static_cast<int>(value), seed);
}

WeighingScaleHandler::WeighingScaleHandler(unsigned short vendorId, unsigned short productId, QObject *parent)
    : QThread(parent), m_vendorId(vendorId), m_productId(productId)
{
}

WeighingScaleHandler::State WeighingScaleHandler::lastStableState() const
{
    return extractState(m_lastStableState);
}

WeighingScaleHandler::State WeighingScaleHandler::instantState() const
{
    return extractState(m_instantState);
}

void WeighingScaleHandler::execOnNextStableWeight(std::function<void(State)> &&handler)
{
    m_stableWaitersLock.lockForWrite();
    m_stableWaiters << handler;
    m_stableWaitersLock.unlock();
}

bool WeighingScaleHandler::isAlive() const
{
    return m_hidHandle;
}

int WeighingScaleHandler::elapsedSinceLastMessage() const
{
    return m_lastSuccessfulRead.elapsed();
}

unsigned short WeighingScaleHandler::vendorId() const
{
    return m_vendorId;
}

unsigned short WeighingScaleHandler::productId() const
{
    return m_productId;
}

void WeighingScaleHandler::stop()
{
    m_stopped = true;
}

void WeighingScaleHandler::run()
{
    int hidResult = hid_init();
    if (hidResult < 0) {
        qCCritical(proofServiceWeighingScaleLog) << "HID API can't be initialized, going down";
        return;
    }

    m_hidHandle = hid_open(m_vendorId, m_productId, NULL);
    if (!m_hidHandle) {
        qCCritical(proofServiceWeighingScaleLog) << "HID device can't be opened, going down";
        return;
    }

    auto stateUpdater = [this](unsigned long long packedState) {
        m_instantState = packedState;
        if (extractStateStatus(packedState) != Status::InMotion) {
            m_lastStableState = packedState;
            m_stableWaitersLock.lockForRead();
            bool waitersExist = m_stableWaiters.size();
            m_stableWaitersLock.unlock();
            if (waitersExist) {
                auto state = extractState(packedState);
                m_stableWaitersLock.lockForWrite();
                while (!m_stableWaiters.isEmpty())
                    m_stableWaiters.takeFirst()(state);
                m_stableWaitersLock.unlock();
            }
        }
    };

    unsigned char data[6];
    memset(data, 0, 6);
    bool warningShown = false;
    m_lastSuccessfulRead.start();
    while (!m_stopped) {
        hidResult = m_hidHandle ? hid_read_timeout(m_hidHandle, data, 6, READ_TIMEOUT) : 0;
        if (hidResult < 6 || data[0] != 3) {
            if (m_lastSuccessfulRead.elapsed() > READ_ERROR_TIMEOUT) {
                if (!warningShown) {
                    qCWarning(proofServiceWeighingScaleLog) << "No readable data from device, setting state to error";
                    warningShown = true;
                }
                stateUpdater(packState(Status::Fault, Unit::Tael, 0, 0));
                hid_close(m_hidHandle);
                QThread::msleep(ERROR_COOLDOWN_TIMEOUT);
                m_hidHandle = hid_open(m_vendorId, m_productId, NULL);
            }
            continue;
        } else if (warningShown) {
            warningShown = false;
            qCDebug(proofServiceWeighingScaleLog) << "Device connection is live again";
        }
        m_lastSuccessfulRead.restart();
        stateUpdater(packState(static_cast<Status>(qMin(data[1], static_cast<unsigned char>(Status::NotInitialized))),
                               static_cast<Unit>(qMin(data[2], static_cast<unsigned char>(Unit::Pound))),
                               static_cast<short>((static_cast<unsigned short>(data[5]) << 8) | static_cast<unsigned short>(data[4])),
                               static_cast<char>(data[3])));
    }
    hid_close(m_hidHandle);
    hid_exit();
}

unsigned long long WeighingScaleHandler::packState(Status status, Unit unit, short weight, char scaleFactor) const
{
    unsigned long long result = static_cast<unsigned short>(weight);
    result |= static_cast<unsigned char>(scaleFactor) << 16;
    result |= static_cast<long long>(unit) << 24;
    result |= static_cast<long long>(status) << 32;
    return result;
}

WeighingScaleHandler::State WeighingScaleHandler::extractState(unsigned long long state) const
{
    double value = static_cast<double>(static_cast<short>(state & 0xFFFF)) * pow(10, static_cast<char>((state >> 16) & 0xFF));
    return State{static_cast<Status>((state >> 32) & 0xFF), static_cast<Unit>((state >> 24) & 0xFF), value};
}

WeighingScaleHandler::Status WeighingScaleHandler::extractStateStatus(unsigned long long state) const
{
    return static_cast<Status>((state >> 32) & 0xFF);
}

QString WeighingScaleHandler::State::stringifiedStatus() const
{
    return STATUSES.value(status, "");
}

QString WeighingScaleHandler::State::stringifiedUnit() const
{
    return UNITS.value(unit, "");
}


weighingscalerestserver.h
#ifndef WEIGHINGSCALERESTSERVER_H
#define WEIGHINGSCALERESTSERVER_H

#include "weighingscalehandler.h"

#include "proofnetwork/abstractrestserver.h"

class WeighingScaleHandler;
class QThread;

class WeighingScaleRestServer : public Proof::AbstractRestServer
{
    Q_OBJECT
public:
    explicit WeighingScaleRestServer(int port);
    ~WeighingScaleRestServer();

protected slots:
    void rest_get_Weight_Instant(QTcpSocket *socket, const QStringList &headers, const QStringList &methodVariableParts,
                                 const QUrlQuery &queryParams, const QByteArray &body);
    void rest_get_Weight_Stable(QTcpSocket *socket, const QStringList &headers, const QStringList &methodVariableParts,
                                const QUrlQuery &queryParams, const QByteArray &body);

protected:
    QMap<QString, QPair<QDateTime, QVariant>> healthStatus(bool quick) const override;

private:
    QJsonObject stateToJson(const WeighingScaleHandler::State &state);
    WeighingScaleHandler *m_handler;
};

#endif // WEIGHINGSCALERESTSERVER_H


weighingscalerestserver.cpp
#include "weighingscalerestserver.h"

#include "proofservice-weighing-scale_global.h"

#include "proofcore/coreapplication.h"
#include "proofcore/settings.h"
#include "proofcore/settingsgroup.h"

#include <QJsonObject>
#include <QJsonDocument>
#include <QTcpSocket>

WeighingScaleRestServer::WeighingScaleRestServer(int port)
    : Proof::AbstractRestServer(port)
{
    QString rawVendor = proofApp->settings()->mainGroup()->value("vendor_id", "0x0b67", Proof::Settings::NotFoundPolicy::Add).toString().trimmed();
    QString rawProduct = proofApp->settings()->mainGroup()->value("product_id", "0x555e", Proof::Settings::NotFoundPolicy::Add).toString().trimmed();
    if (rawVendor.startsWith("0x", Qt::CaseInsensitive))
        rawVendor.remove(0, 2);
    if (rawProduct.startsWith("0x", Qt::CaseInsensitive))
        rawProduct.remove(0, 2);
    m_handler = new WeighingScaleHandler(rawVendor.toUShort(nullptr, 16), rawProduct.toUShort(nullptr, 16));
    m_handler->moveToThread(qApp->thread());
    m_handler->start();
}

WeighingScaleRestServer::~WeighingScaleRestServer()
{
    m_handler->stop();
    m_handler->wait(1000);
    delete m_handler;
}

void WeighingScaleRestServer::rest_get_Weight_Instant(QTcpSocket *socket, const QStringList &, const QStringList &, const QUrlQuery &, const QByteArray &)
{
    sendAnswer(socket, QJsonDocument(stateToJson(m_handler->instantState())).toJson(QJsonDocument::Compact), "application/json", 200, "OK");
}

void WeighingScaleRestServer::rest_get_Weight_Stable(QTcpSocket *socket, const QStringList &, const QStringList &, const QUrlQuery &queryParams, const QByteArray &)
{
    bool immediate = queryParams.hasQueryItem("immediate");
    if (immediate) {
        sendAnswer(socket, QJsonDocument(stateToJson(m_handler->lastStableState())).toJson(QJsonDocument::Compact), "application/json", 200, "OK");
    } else {
        m_handler->execOnNextStableWeight([this, socket](WeighingScaleHandler::State state) {
            sendAnswer(socket, QJsonDocument(stateToJson(state)).toJson(QJsonDocument::Compact), "application/json", 200, "OK");
        });
    }
}

QMap<QString, QPair<QDateTime, QVariant>> WeighingScaleRestServer::healthStatus(bool) const
{
    QDateTime current = QDateTime::currentDateTimeUtc();
    return QMap<QString, QPair<QDateTime, QVariant> > {
        {QStringLiteral("device_vendor_id"), {current, QString("0x%1").arg(m_handler->vendorId(), 4, 16, QChar('0'))}},
        {QStringLiteral("device_product_id"), {current, QString("0x%1").arg(m_handler->productId(), 4, 16, QChar('0'))}},
        {QStringLiteral("device_alive"), {current, m_handler->isAlive()}},
        {QStringLiteral("msecs_since_last_read_from_device"), {current, m_handler->elapsedSinceLastMessage()}}
    };
}

QJsonObject WeighingScaleRestServer::stateToJson(const WeighingScaleHandler::State &state)
{
    QJsonObject result;
    result.insert("weight", state.weight);
    result.insert("units", state.stringifiedUnit());
    result.insert("status", state.stringifiedStatus());
    return result;
}
